Add refresh token documentation specific to the OIE SDK and widget (#2291)
okta/okta-developer-docs: packages/@okta/vuepress-site/docs/guides/oie-embedded-common-refresh-tokens/main/index.md

<ApiLifecycle access="ie" /><br>

<StackSelector class="cleaner-selector"/>

## Summary

Access tokens often have limited lifetimes. Allowing access
tokens to expire limits their usefulness in case they are discovered
by an attacker. In order for your app to continue to use the
appropriate resources when an access token expires, it can refresh
them without user intervention through the use of a refresh token.

All the SDKs expose functionality that allow you
to obtain access, ID, and refresh tokens. Depending on the SDK (Swift,
Javascript, and so on), you have varying degrees of convenience
methods and other functionality that provide you with built-in support
for token refresh, auto-renewal, and storage.  Minimally,
all the SDKs allow you to obtain the refresh token and call the authorization
server's token endpoint to renew the access token.

<StackSelector snippet="refreshusingthesdk" noSelector />

## Option 2: Refresh using the OAuth token endpoint

You can refresh access and ID tokens using the
[`/token`](https://developer.okta.com/docs/reference/api/oidc/#token)
endpoint with the `grant_type` set to `refresh_token`.  Before calling this endpoint,

> obtain the refresh token from the SDK and ensure you have included

Reviewer: ```suggestion
obtain the refresh token from the SDK and ensure that you have included
```

`offline_access` as a scope in the SDK configurations. For further details on
access token refresh with this endpoint, see
[Use a refresh token](/docs/guides/refresh-tokens/use-refresh-token/).

<StackSelector snippet="refreshendpointrequest" noSelector />

### Native and Spa applications

Unlike web applications, native applications and single-page applications
don't use client secrets. As a result, token refresh requests for these
applications omit the `Authorization` header and instead include the
`client_id` query parameter.

Example

```http
POST /oauth2/default/v1/token HTTP/1.1
Accept: application/json
Content-Type: application/x-www-form-urlencoded

grant_type=refresh_token
redirect_uri=com.embeddedauth://callback
scope=offline_access openid profile
refresh_token=03_hBtVj-Hk0Mxo9TPSdl7TLkxQioKqQEzud3ldqHqs
client_id=0oa94el1z4nUDxx0z5d7
```

### Response

> An example response is shown below:

Reviewer: ```suggestion

```

```json
{
    "token_type": "Bearer",
    "expires_in": 3600,
    "access_token": "eyJraWQiOiJoQkZNR...",
    "scope": "offline_access openid profile",
    "refresh_token": "HRzOBfj1A1g6akWqNHfCE-KX-9NASmnFqhRFOt_rEdc",
    "id_token": "eyJraWQiOiJoQkZN..."
}
```

## Get token info using introspect endpoint

> To learn more information about a tokens (access or Id) including whether

Reviewer: ```suggestion
To learn more information about tokens (access or Id), including whether
```

they have expired, use the
[`/introspect`](https://developer.okta.com/docs/reference/api/oidc/#introspect)
endpoint.

<StackSelector snippet="introspectendpointrequest" noSelector />

> ### Native and Spa applications

Reviewer: ```suggestion
### Native and SPA applications
```

> Unlike web applications, native applications and single-page applications don't

Reviewer: ```suggestion
Unlike web applications, native applications and single-page applications (SPAs) don't
```

use client secrets. Refresh requests for these applications omit the `Authorization`
header and include the `client_id` query parameter.

Example

```http
POST /oauth2/default/v1/introspect HTTP/1.1
Accept: application/json
Content-Type: application/x-www-form-urlencoded

token=eyJraWQiOiJoQk...
client_id=0oa14dl1z4nUJxx0z5d7
token_type_hint=access_token
```

### Response

> An example response is shown below:

Reviewer: ```suggestion

```

---

> title: Refresh access and id tokens

Reviewer: ```suggestion
title: Refresh access and ID tokens
```

---
<div class="oie-embedded-sdk">

<ApiLifecycle access="ie" /><br>

<StackSelector class="cleaner-selector"/>

## Summary

Access tokens often have limited lifetimes. Allowing access

> tokens to expire limits there usefulness in case they are discovered

Reviewer 2: ```suggestion
tokens to expire limits their usefulness in case they are discovered
```

by an attacker. In order for your app to continue to use the
appropriate resources with an expired access token, it can refresh
them without user intervention through the use of a refresh token.

> All the SDK's expose functionality that allow you

Reviewer: ```suggestion
All the SDKs expose functionality that allow you
```

to obtain these tokens. Depending on the SDK (Swift, Javascript,

>  etc), you have varying degrees of convenience

Reviewer: ```suggestion
 and so on), you have varying degrees of convenience
```

methods and other functionality that provide you with built-in support
for token refresh, auto-renewal, and storage.  Minimally,

> all the SDK's allow you to obtain the refresh token and call the authorization

Reviewer: ```suggestion
all the SDKs allow you to obtain the refresh token and call the authorization
```

server's token endpoint to renew the access token.

<StackSelector snippet="refreshusingthesdk" noSelector />

## Refresh using the OAuth token endpoint

> Access and id tokens can be refreshed using OAuth token endpoint

Reviewer: ```suggestion
You can refresh access and ID tokens using the OAuth token endpoint
```

described in the OAuth
[specification](https://datatracker.ietf.org/doc/html/rfc6749#section-5.1).

> Before calling this endpoint, obtain the refresh token from the SDK. Once

Reviewer: ```suggestion
Before calling this endpoint, obtain the refresh token from the SDK. After you retrieve the refresh token, 
```

> retrieved, create an HTTP Post request to the authorization server's `token` endpoint.

Reviewer: ```suggestion
create an HTTP Post request to the authorization server's `token` endpoint.
```

> See example below.

Reviewer: ```suggestion
Example
```

```http
POST /oauth2/default/v1/token HTTP/1.1
Accept: application/json
Content-Type: application/x-www-form-urlencoded
Authorization: Basic MG9hMTJncG5qZnpvdllTbk41ZDc6UG9zWGpjNmw0WHF5ZDBhek03cjF0SnhyMS1LWHdWYmNFaDk0Q0FDNA==

grant_type: refresh_token
redirect_uri: http://localhost:8080
scope: offline_access openid profile
refresh_token: 03_hBtVj-Hk0Mxo9TPSdl7TLkxQioKqQEzud3ldqHqs
```

> Note the use of the `refresh_token` `grant_type` value which indicates

Reviewer: ```suggestion
Note the use of the `refresh_token` as the `grant_type` value, which indicates that
```
